tracer/src/Datadog.Trace/Debugger/SpanCodeOrigin/SpanCodeOriginManager.cs

```diff
@@ -5,94 +5,219 @@
 using Datadog.Trace.VendoredMicrosoftCode.System.Buffers;
+using Datadog.Trace.VendoredMicrosoftCode.System.Collections.Immutable;
 
 namespace Datadog.Trace.Debugger.SpanCodeOrigin
 {
     internal class SpanCodeOriginManager
     {
-        private const string CodeOriginTag = "_dd.code_origin";
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(SpanCodeOriginManager));
 
-        private const string FramesPrefix = "frames";
+        private readonly ConcurrentDictionary<Assembly, AssemblyPdbInfo?> _assemblyPdbCache = new();
 
-        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(SpanCodeOriginManager));
+        private readonly CodeOriginTags _tags;
+
+        private readonly int _maxUserFrames;
 
-        private readonly DebuggerSettings _settings = LiveDebugger.Instance?.Settings ?? DebuggerSettings.FromDefaultSource();
+        private readonly bool _isCodeOriginEnabled;
+
+        private readonly ImmutableHashSet<string> _thirdPartyDetectionExcludes;
+
+        private readonly ImmutableHashSet<string> _thirdPartyDetectionIncludes;
+
+        private SpanCodeOriginManager()
+        {
+            _tags = new CodeOriginTags(_maxUserFrames);
+            var settings = LiveDebugger.Instance?.Settings ?? DebuggerSettings.FromDefaultSource();
+            _maxUserFrames = settings.CodeOriginMaxUserFrames;
+            _isCodeOriginEnabled = settings.CodeOriginForSpansEnabled;
+            _thirdPartyDetectionExcludes = settings.ThirdPartyDetectionExcludes;
+            _thirdPartyDetectionIncludes = settings.ThirdPartyDetectionIncludes;
+        }
 
         internal static SpanCodeOriginManager Instance { get; } = new();
 
-        internal void SetCodeOrigin(Span? span)
+        internal void SetCodeOriginForExitSpan(Span? span)
+        {
+            if (span == null || !_isCodeOriginEnabled)
+            {
+                return;
+            }
+
+            if (span.Tags is WebTags { SpanKind: SpanKinds.Server })
+            {
+                // entry span
+                Log.Debug("Skipping span {SpanID}, we will add entry span code origin for it later. Resource: {ResourceName}, Operation: {OperationName}", span.SpanId, span.ResourceName, span.OperationName);
+                return;
+            }
+
+            if (span.GetTag(_tags.Type) != null)
+            {
+                Log.Debug("Span {SpanID} has already code origin tags. Resource: {ResourceName}, Operation: {OperationName}", span.SpanId, span.ResourceName, span.OperationName);
+                return;
+            }
+
+            AddExitSpanTags(span);
+        }
+
+        internal void SetCodeOriginForEntrySpan(Span? span, Type? type, MethodInfo? method)
         {
-            if (span == null || !_settings.CodeOriginForSpansEnabled)
+            if (span == null ||
+                type == null ||
+                method == null ||
+                !this._isCodeOriginEnabled)
```

Comment:
Maybe `IsCodeOriginEnabled` should be lifted up to the same level where `shouldTrace`, `shouldSecure` and `shouldUseIast` (in `AspNetCoreDiagnosticObserver.OnMvcBeforeAction`) for consistency? The other benefit of doing so will be to skip `TryDuckCast` calls that take place in hot path